consensus/src/quorum_store/utils.rs

```diff
@@ -188,159 +215,162 @@ impl ProofQueue {
         max_txns: u64,
         max_bytes: u64,
         return_non_full: bool,
     ) -> Vec<ProofOfStore> {
-        let num_expired = self
-            .digest_queue
-            .iter()
-            .take_while(|(_, expiration_time)| *expiration_time < current_block_timestamp)
-            .count();
-        let mut num_expired_but_not_committed = 0;
-        for (digest, expiration_time) in self.digest_queue.drain(0..num_expired) {
-            if self
-                .digest_proof
-                .get(&digest)
-                .expect("Entry for unexpired digest must exist")
-                .is_some()
-            {
-                // non-committed proof that is expired
-                num_expired_but_not_committed += 1;
-                if expiration_time < current_block_timestamp {
-                    counters::GAP_BETWEEN_BATCH_EXPIRATION_AND_CURRENT_TIME_WHEN_PULL_PROOFS
-                        .observe((current_block_timestamp - expiration_time) as f64);
-                }
-            }
-            claims::assert_some!(self.digest_proof.remove(&digest));
-            self.digest_insertion_time.remove(&digest);
-        }
-
-        let mut ret = Vec::new();
+        let mut ret = vec![];
         let mut cur_bytes = 0;
         let mut cur_txns = 0;
-        let initial_size = self.digest_queue.len();
-        let mut size = self.digest_queue.len();
+        let mut excluded_txns = 0;
         let mut full = false;
 
-        for (digest, expiration) in self
-            .digest_queue
-            .iter()
-            .filter(|(digest, _)| !excluded_proofs.contains(digest))
-        {
-            if let Some(proof) = self
-                .digest_proof
-                .get(digest)
-                .expect("Entry for unexpired digest must exist")
-            {
-                if *expiration >= current_block_timestamp {
-                    // non-committed proof that has not expired
-                    cur_bytes += proof.num_bytes();
-                    cur_txns += proof.num_txns();
-                    if cur_bytes > max_bytes || cur_txns > max_txns {
-                        // Exceeded the limit for requested bytes or number of transactions.
-                        full = true;
-                        break;
-                    }
-                    ret.push(proof.clone());
-                    if let Some(insertion_time) = self.digest_insertion_time.get(digest) {
+        let mut author_to_num_remaining = HashMap::new();
+        for (author, batches) in self.author_to_batches.iter() {
+            author_to_num_remaining.insert(*author, batches.len());
+        }
+
+        'outer: while !author_to_num_remaining.is_empty() {
+            let shuffled_peers: Vec<_> = {
+                let mut peers: Vec<_> = author_to_num_remaining.keys().cloned().collect();
+                peers.shuffle(&mut thread_rng());
+                peers
+            };
+
+            for peer in shuffled_peers {
```

Comment:
Hmm, I thought by fairness, we would try to represent many peers' batches in a block, but looks like we are only achieving fairness among peers between blocks when all the queues are equally full and can meet the block size.